File: Part-2-Regression/Section 9 - Support Vector Regression (SVR)/R/SVR.R
# importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

# Splitting the dataset into training and test sets
#library('caTools')
#set.seed(123)
#split = sample.split(dataset$Purchased , SplitRatio = 0.8)
#training_set = subset(dataset , split == TRUE)
#test_set = subset(dataset , split == FALSE)

# Feature scaling
# in R we should exclude the categorical features 
# if we include it this will cause an error
# training_set[ ,2:3] = scale(training_set[ ,2:3])
# test_set[ ,2:3] = scale(test_set[ ,2:3])


# fitting  SVR to the dataset
#install.packages('e1071')
library('e1071')
regressor = svm(formula = Salary ~.,
                data = dataset,
                type = 'eps-regression')

# predicting a new result with SVR

y_pred = predict(regressor , data.frame(Level = 6.5))

# visualizing the SVR results

ggplot()+
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = dataset$Level , y = predict(regressor , newdata = dataset)),
            colour = 'blue')+
  ggtitle('Position Levels vs Salary (SVR)')+
  xlab('Position Levels')+
  ylab('Salary')
# visualizing the SVR results with higher resolution 
x_grid = seq(min(dataset$Level),max(dataset$Level),0.01)
ggplot()+
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = x_grid , y = predict(regressor , newdata = data.frame(Level = x_grid))),
            colour = 'blue')+
  ggtitle('Position Levels vs Salary (SVR)')+
  xlab('Position Levels')+
  ylab('Salary')

File: Part-2-Regression/Section 9 - Support Vector Regression (SVR)/Python/SVR.py
# -*- coding: utf-8 -*-
"""
Created on Wed Sep 16 00:44:05 2020

@author: eman
"""
# importing libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd 

# importing dataset

dataset = pd.read_csv('Position_Salaries.csv')
# Features values
x = dataset.iloc[:,1:2].values
# Target values
y = dataset.iloc[:,2:3].values
# 2:3 because fit_transform method needs an array and we need to define y as an array contains the column number 2
# when we write 2:3 the first number is included and the second number is excluded so y contains only one column which is number 2
"""# split the dataset into train and test dataset
# random state make the result be the same as the other people if it setted like them
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=0)
"""
# we don't split the data bescause the data is very small so we need all the data to make accurate prediction
# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc_x = StandardScaler()
sc_y = StandardScaler()
x= sc_x.fit_transform(x)
y= sc_y.fit_transform(y)
# we have to make feature scaling because svr didn't make feature scaling by itself
# fitting SVR to the dataset


from sklearn.svm import SVR
regressor = SVR(kernel = 'rbf')
regressor.fit(x,y)

# predicting a new result with SVR

y_pred = sc_y.inverse_transform(regressor.predict(sc_x.transform(np.array([[6.5]]))))

# visualizing the SVR of the dataset
plt.scatter(x, y, color = 'red')
plt.plot(x,regressor.predict(x),color = 'blue')
plt.title('Salary vs Position Levels (SVR)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

# visualizing the SVR of the dataset (for higher resolution and smoother curve)

x_grid = np.arange(min(x),max(x), 0.1)
x_grid = x_grid.reshape((len(x_grid),1))
plt.scatter(x, y, color = 'red')
plt.plot(x_grid,regressor.predict(x_grid),color = 'blue')
plt.title('Salary vs Position Levels (SVR)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

